compiler/internal/typeparams/collect.go

```diff
@@ -0,0 +1,266 @@
+package typeparams
+
+import (
+	"fmt"
+	"go/ast"
+	"go/types"
+
+	"github.com/gopherjs/gopherjs/compiler/internal/typeparams/subst"
+	"github.com/gopherjs/gopherjs/compiler/typesutil"
+	"golang.org/x/exp/typeparams"
+)
+
+// Resolver translates types defined in terms of type parameters into concrete
+// types, given a mapping from type params to type arguments.
+type Resolver struct {
+	subster *subst.Subster
+	selMemo map[typesutil.Selection]typesutil.Selection
+}
+
+// NewResolver creates a new Resolver with tParams entries mapping to tArgs
+// entries with the same index.
+func NewResolver(tc *types.Context, tParams []*types.TypeParam, tArgs []types.Type) *Resolver {
+	r := &Resolver{
+		subster: subst.New(tc, tParams, tArgs),
+		selMemo: map[typesutil.Selection]typesutil.Selection{},
+	}
+	if len(tParams) != len(tArgs) {
```

Comment:
Does it make sense to do this check/panic before passing the checked arguments to `subst.New`?

Edit: Now that I look at subst.New, it also seems to panic on the same condition. so maybe we can just eliminate this check/panic entirely?

Reply:
Ah, yes, this is a holdover from an earlier revision that used unexported go/types function that did not have the check. Removed.